source/connection.c

```diff
@@ -680,23 +700,37 @@ struct aws_http_server *aws_http_server_new(const struct aws_http_server_options
         .incoming_callback = s_server_bootstrap_on_accept_channel_setup,
         .shutdown_callback = s_server_bootstrap_on_accept_channel_shutdown,
         .destroy_callback = s_server_bootstrap_on_server_listener_destroy,
+        .setup_callback = s_server_bootstrap_on_server_listener_setup,
         .host_name = options->endpoint->address,
         .port = options->endpoint->port,
         .user_data = server,
     };
 
-    server->socket = aws_server_bootstrap_new_socket_listener(&bootstrap_options);
+    int listen_error = AWS_OP_SUCCESS;
 
+    /*
+     * WARNING & TODO!!!!
+     * aws_server_bootstrap_new_socket_listener has async callback, we would block here waiting for
+     * setup complete. aws-c-http library need to be updated with a proper async API.
+     */
+    server->socket = aws_server_bootstrap_new_socket_listener(&bootstrap_options);
+    // if server setup properly, waiting for setup callback
+    if (server->socket) {
+        aws_future_void_wait(server->setup_future, UINT64_MAX /*timeout*/);
+        listen_error = aws_future_void_get_error(server->setup_future);
```

Comment:
in this path, `aws_raise_error()` is never called, so the aws_http_server_new() is going to return NULL, but if the user calls aws_last_error() to find out why they'll get 0 or whatever stale old error was on the stack